### ISSUE TITLE
Bug with IIS 7.0: Invalid attribute startMode for the pool

### ISSUE THREAD

**Issue by olivierlemasle** [opscode-cookbooks/iis#185]
I'm using iis cookbook 4.1.1 on Windows 2008 (R1), with IIS 7.0.

When using:

```ruby
iis_pool 'DefaultAppPool' do
  runtime_version "4.0"
  action :config
end
```
I get the following error:

    ---- Begin output of C:\Windows\System32\inetsrv\appcmd.exe set config /section:applicationPools "/[name='DefaultAppPool'].startMode:OnDemand" "/[name='DefaultAppPool'].managedRuntimeVersion:v4.0" "/[name='DefaultAppPool'].processModel.logonType:LogonBatch" "/[name='DefaultAppPool'].recycling.logEventOnRecycle:PrivateMemory,Memory,Schedule,Requests,Time,ConfigChange,OnDemand,IsapiUnhealthy" ----
    STDOUT: ERROR ( message:Unknown attribute "[name='DefaultAppPool'].startMode".  Replace with -? for help. )

The issue is caused by the fact that the  `startMode` is not defined for IIS 7.0 (it has been added in IIS 7.5: see http://www.iis.net/configreference/system.applicationhost/applicationpools/applicationpooldefaults)

I did not specify any `start_mode` but OnDemand is the default one.
There is no startMode in the current configuration of my pool, therefore it is considered as an update (cf line 
https://github.com/opscode-cookbooks/iis/blob/master/providers/pool.rb#L137).

The parameter values should be updated only if they're already defined.

**Comment by EasyAsABC123** [comment]
The issue with that is that they are auto defined when you create the iis pool object since they have default values...

Solutions:
1. Check iis version and add version check to each parameter that doesn't exist in other iis versions 
2. Only update if new and not default, this one is tricky since if you changed the default to something else then want to change it back you'd have to manually specify it rather than simply remove your updated value

I think solution 1 is the proper solution, since it will solve the issue you are mentioning and allow for more properties to be added.

Thanks for finding this bug!

Justin Schuhmann

**Comment by EasyAsABC123** [comment]
please test this version: https://github.com/opscode-cookbooks/iis/tree/enhancement-iis_version_detection

You can download it here: https://github.com/opscode-cookbooks/iis/archive/enhancement-iis_version_detection.zip

Let me know and thanks @olivierlemasle

**Comment by olivierlemasle** [comment]
Thanks, @EasyAsABC123.

But there is an error with this branch:
```
==> web: ================================================================================
==> web: Recipe Compile Error in C:\tmp\vagrant-chef\65abd8a7ac6dcd783e8e13e8ee74088a\cookbooks\iis\libraries\helper.rb
==> web: ================================================================================
==> web: 
==> web: SyntaxError
==> web: -----------
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: syntax error, unexpected ->, expecting keyword_end
==> web:         $HKLM->Open('SOFTWARE\Microsoft\Inet...
==> web:                ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: formal argument cannot be a constant
==> web:         $HKLM->Open('SOFTWARE\Microsoft\InetStp'...
==> web:                    ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: syntax error, unexpected ')', expecting keyword_end
==> web: ...n('SOFTWARE\Microsoft\InetStp')->GetValue('VersionString').l...
==> web: ...                               ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: formal argument cannot be a constant
==> web: ...E\Microsoft\InetStp')->GetValue('VersionString').lstrip('Ver...
==> web: ...                               ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: syntax error, unexpected ')', expecting keyword_end
==> web: ...tp')->GetValue('VersionString').lstrip('Version ');
==> web: ...                               ^
==> web: 
==> web: Cookbook Trace:
==> web: ---------------
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:191:in `load'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:191:in `block in load_libraries_from_cookbook'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:188:in `each'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:188:in `load_libraries_from_cookbook'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:99:in `block in compile_libraries'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:98:in `each'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:98:in `compile_libraries'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:71:in `compile'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context.rb:96:in `load'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/policy_builder/expand_node_object.rb:87:in `setup_run_context'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/client.rb:256:in `setup_run_context'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/client.rb:454:in `run'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:253:in `run_with_graceful_exit_option'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:230:in `block in run_chef_client'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/local_mode.rb:39:in `with_server_connectivity'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:213:in `run_chef_client'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application/solo.rb:280:in `block in interval_run_chef_client'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application/solo.rb:269:in `loop'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application/solo.rb:269:in `interval_run_chef_client'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application/solo.rb:247:in `run_application'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:60:in `run'
==> web:   C:/opscode/chef/embedded/apps/chef/bin/chef-solo:25:in `<top (required)>'
==> web:   C:/opscode/chef/bin/chef-solo:63:in `load'
==> web:   C:/opscode/chef/bin/chef-solo:63:in `<main>'
==> web: 
==> web: Relevant File Content:
==> web: ----------------------
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:
==> web: 
==> web: 184:          raise
==> web: 185:        end
==> web: 186:  
==> web: 187:        def load_libraries_from_cookbook(cookbook_name)
==> web: 188:          files_in_cookbook_by_segment(cookbook_name, :libraries).each do |filename|
==> web: 189:            begin
==> web: 190:              Chef::Log.debug("Loading cookbook #{cookbook_name}'s library file: #{filename}")
==> web: 191>>             Kernel.load(filename)
==> web: 192:              @events.library_file_loaded(filename)
==> web: 193:            rescue Exception => e
==> web: 194:              @events.library_file_load_failed(filename, e)
==> web: 195:              raise
==> web: 196:            end
==> web: 197:          end
==> web: 198:        end
==> web: 199:  
==> web: 200:        def load_lwrps_from_cookbook(cookbook_name)
==> web: 
==> web: [2015-05-27T11:01:35-07:00] DEBUG: Re-raising exception: SyntaxError - C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: syntax error, unexpected ->, expecting keyword_end
==> web:         $HKLM->Open('SOFTWARE\Microsoft\Inet...
==> web:                ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: formal argument cannot be a constant
==> web:         $HKLM->Open('SOFTWARE\Microsoft\InetStp'...
==> web:                    ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: syntax error, unexpected ')', expecting keyword_end
==> web: ...n('SOFTWARE\Microsoft\InetStp')->GetValue('VersionString').l...
==> web: ...                               ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: formal argument cannot be a constant
==> web: ...E\Microsoft\InetStp')->GetValue('VersionString').lstrip('Ver...
==> web: ...                               ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: syntax error, unexpected ')', expecting keyword_end
==> web: ...tp')->GetValue('VersionString').lstrip('Version ');
==> web: ...                               ^
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:191:in `load'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:191:in `block in load_libraries_from_cookbook'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:188:in `each'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:188:in `load_libraries_from_cookbook'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:99:in `block in compile_libraries'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:98:in `each'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:98:in `compile_libraries'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:71:in `compile'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/run_context.rb:96:in `load'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/policy_builder/expand_node_object.rb:87:in `setup_run_context'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/client.rb:256:in `setup_run_context'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/client.rb:454:in `run'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:253:in `run_with_graceful_exit_option'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:230:in `block in run_chef_client'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/local_mode.rb:39:in `with_server_connectivity'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:213:in `run_chef_client'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application/solo.rb:280:in `block in interval_run_chef_client'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application/solo.rb:269:in `loop'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application/solo.rb:269:in `interval_run_chef_client'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application/solo.rb:247:in `run_application'
==> web:   C:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:60:in `run'
==> web:   C:/opscode/chef/embedded/apps/chef/bin/chef-solo:25:in `<top (required)>'
==> web:   C:/opscode/chef/bin/chef-solo:63:in `load'
==> web:   C:/opscode/chef/bin/chef-solo:63:in `<main>'
==> web: [2015-05-27T11:01:35-07:00] ERROR: Running exception handlers
==> web: [2015-05-27T11:01:35-07:00] ERROR: Exception handlers complete
==> web: [2015-05-27T11:01:35-07:00] FATAL: Stacktrace dumped to C:/var/chef/cache/chef-stacktrace.out
==> web: [2015-05-27T11:01:35-07:00] DEBUG: SyntaxError: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: syntax error, unexpected ->, expecting keyword_end
==> web:         $HKLM->Open('SOFTWARE\Microsoft\Inet...
==> web:                ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: formal argument cannot be a constant
==> web:         $HKLM->Open('SOFTWARE\Microsoft\InetStp'...
==> web:                    ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: syntax error, unexpected ')', expecting keyword_end
==> web: ...n('SOFTWARE\Microsoft\InetStp')->GetValue('VersionString').l...
==> web: ...                               ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: formal argument cannot be a constant
==> web: ...E\Microsoft\InetStp')->GetValue('VersionString').lstrip('Ver...
==> web: ...                               ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: syntax error, unexpected ')', expecting keyword_end
==> web: ...tp')->GetValue('VersionString').lstrip('Version ');
==> web: ...                               ^
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:191:in `load'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:191:in `block in load_libraries_from_cookbook'
==> web: 
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:188:in `each'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:188:in `load_libraries_from_cookbook'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:99:in `block in compile_libraries'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:98:in `each'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:98:in `compile_libraries'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:71:in `compile'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/run_context.rb:96:in `load'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/policy_builder/expand_node_object.rb:87:in `setup_run_context'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/client.rb:256:in `setup_run_context'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/client.rb:454:in `run'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:253:in `run_with_graceful_exit_option'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:230:in `block in run_chef_client'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/local_mode.rb:39:in `with_server_connectivity'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:213:in `run_chef_client'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/application/solo.rb:280:in `block in interval_run_chef_client'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/application/solo.rb:269:in `loop'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/application/solo.rb:269:in `interval_run_chef_client'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/application/solo.rb:247:in `run_application'
==> web: C:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:60:in `run'
==> web: C:/opscode/chef/embedded/apps/chef/bin/chef-solo:25:in `<top (required)>'
==> web: C:/opscode/chef/bin/chef-solo:63:in `load'
==> web: C:/opscode/chef/bin/chef-solo:63:in `<main>'
==> web: [2015-05-27T11:01:35-07:00] FATAL: SyntaxError: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: syntax error, unexpected ->, expecting keyword_end
==> web:         $HKLM->Open('SOFTWARE\Microsoft\Inet...
==> web:                ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: formal argument cannot be a constant
==> web:         $HKLM->Open('SOFTWARE\Microsoft\InetStp'...
==> web:                    ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: syntax error, unexpected ')', expecting keyword_end
==> web: ...n('SOFTWARE\Microsoft\InetStp')->GetValue('VersionString').l...
==> web: ...                               ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: formal argument cannot be a constant
==> web: ...E\Microsoft\InetStp')->GetValue('VersionString').lstrip('Ver...
==> web: ...                               ^
==> web: C:/tmp/vagrant-chef/65abd8a7ac6dcd783e8e13e8ee74088a/cookbooks/iis/libraries/helper.rb:167: syntax error, unexpected ')', expecting keyword_end
==> web: ...tp')->GetValue('VersionString').lstrip('Version ');
==> web: ...                               ^
Chef never successfully completed! Any errors should be visible in the
output above. Please fix your recipes so that they properly complete.
```

**Comment by EasyAsABC123** [comment]
No worries tonight I'll run tests and validate it. I might even have a 2008r1 server install somewhere.  Can add it to my test kitchen, I apologize for the error it's a minor one that I should of caught but I don't have a machine to test with always...

Justin Schuhmann

**Comment by olivierlemasle** [comment]
No problem!
I'm new to both Ruby and Chef; otherwise, I would have hepled with pleasure.
But I will test the final fix!

Thanks

**Comment by EasyAsABC123** [comment]
Never a problem! If you ever want to get a better understanding of how this all works is be happy to walk you through quickly, as this cookbook is a light weight resource provider.  Nothing too complex about it. :). Let me know we can use IRC or Google hangouts

Justin Schuhmann

**Comment by EasyAsABC123** [comment]
@olivierlemasle This is updated and tested, please download latest and validate it works for your server.  I didn't have a Windows Server 2008 box so i'm hoping you can validate it fixes your specific issue.

**Comment by olivierlemasle** [comment]
I will test in the weekend, thanks.

**Comment by olivierlemasle** [comment]
@EasyAsABC123, I've finally tested on Windows Server 2008 R1.

There was two remaining problems:

1. Like `startMode`, parameter `logonType` (on [Process Model](http://www.iis.net/configreference/system.applicationhost/applicationpools/add/processmodel)), was introduced in IIS 7.5.

2. `processModel.identityType` was already defined in IIS, but its value ApplicationPoolIdentity (which is the default one since 7.5) was only introduced in IIS 7.5.

        ERROR ( message:Unknown attribute "identityType".. Reason: Enum must be one of LocalSystem, LocalService, NetworkService, SpecificUser

Cf. http://www.iis.net/configreference/system.applicationhost/applicationpools/applicationpooldefaults
and http://www.iis.net/configreference/system.applicationhost/applicationpools/add/processmodel

I've added the condition IIS > 7 for these two fields, and it now works.
Thanks!

**Comment by EasyAsABC123** [comment]
Sounds good, I'm going to likely do one more refactor then it should be good to go

Justin Schuhmann

**Comment by EasyAsABC123** [comment]
Everything is good to go, looking into switching over to annotations in ruby so that a lot can be conglomerated into the same attribute (or bound to).  But this is good to go, just need @ksubrama to do a quick code review to get it all pulled to master.

**Comment by EasyAsABC123** [comment]
resolved by #195
